samples/go/csv/read.go

```diff
@@ -131,24 +168,40 @@ func ReadToMap(r *csv.Reader, structName string, headersRaw []string, pkIdx int,
 			panic(err)
 		}
 
-		var pk types.Value
-		fields := make(types.ValueSlice, len(headersRaw))
-		for i, v := range row {
-			if i < len(headersRaw) {
-				fieldOrigIndex := fieldOrder[i]
-				fields[fieldOrigIndex], err = StringToValue(v, kindMap[fieldOrigIndex])
-				if i == pkIdx {
-					pk = fields[fieldOrigIndex]
-				}
-				if err != nil {
-					d.Chk.Fail(fmt.Sprintf("Error parsing value for column '%s': %s", headersRaw[i], err))
-				}
+		fields := readFieldsFromRow(row, headersRaw, fieldOrder, kindMap)
+		rowStruct := types.NewStructWithType(t, fields)
+
+		// needed to allow recursive calls to encloseInMap
+		var encloseInMapFunc func(m types.Map, keyLevel int) types.Map
+		encloseInMap := func(m types.Map, keyLevel int) types.Map {
+			fieldOrigIndex := fieldOrder[pkIndices[keyLevel]]
+			key := fields[fieldOrigIndex]
+
+			// at end of our indices, set the final key to point to this row
+			if keyLevel == len(pkIndices)-1 {
+				return m.Set(key, rowStruct)
+			}
+
+			// not at end of our indices, determine if we already have a map
+			// created for the next level and use it if so, otherwise create it
+			var subMap types.Map
+			if n, ok := m.MaybeGet(key); !ok {
+				subMap = types.NewMap()
+			} else {
+				subMap = n.(types.Map)
 			}
+			return m.Set(key, encloseInMapFunc(subMap, keyLevel+1))
 		}
-		kvChan <- pk
-		kvChan <- types.NewStructWithType(t, fields)
+
+		encloseInMapFunc = encloseInMap
+		topMap = encloseInMap(topMap, 0)
 	}
 
+	topMap.IterAll(func(k types.Value, v types.Value) {
```

Comment:
This is just taking things out of the map you've already build in memory, streaming them to the db, and then building another map. It doesn't seem useful.